server/src/main/java/org/elasticsearch/readiness/ReadinessService.java

```diff
@@ -0,0 +1,227 @@
+import org.elasticsearch.cluster.ClusterChangedEvent;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.ClusterStateListener;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.component.AbstractLifecycleComponent;
+import org.elasticsearch.common.io.Channels;
+import org.elasticsearch.common.network.NetworkAddress;
+import org.elasticsearch.common.settings.Setting;
+import org.elasticsearch.common.transport.BoundTransportAddress;
+import org.elasticsearch.common.transport.PortsRange;
+import org.elasticsearch.common.transport.TransportAddress;
+import org.elasticsearch.common.util.concurrent.EsExecutors;
+import org.elasticsearch.common.util.concurrent.ThreadContext;
+import org.elasticsearch.env.Environment;
+import org.elasticsearch.http.HttpServerTransport;
+import org.elasticsearch.shutdown.PluginShutdownService;
+import org.elasticsearch.transport.BindTransportException;
+
+import java.io.IOException;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.net.StandardProtocolFamily;
+import java.nio.channels.ServerSocketChannel;
+import java.nio.channels.SocketChannel;
+import java.nio.charset.StandardCharsets;
+import java.security.AccessController;
+import java.security.PrivilegedAction;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.function.Function;
+
+public class ReadinessService extends AbstractLifecycleComponent implements ClusterStateListener {
+    private static final Logger logger = LogManager.getLogger(ReadinessService.class);
+    private static final int RESPONSE_TIMEOUT_MILLIS = 1_000;
+
+    private final Environment environment;
+    private final ExecutorService workerExecutor;
+
+    private volatile ServerSocketChannel serverChannel;
+
+    private volatile boolean ready = false;
+    private volatile BoundTransportAddress boundAddress;
+
+    private final ReadWriteLock closeLock = new ReentrantReadWriteLock();
+
+    public static final Setting<String> PORT = new Setting<>(
+        "readiness.port",
+        "9400-9500",
+        Function.identity(),
+        Setting.Property.NodeScope
+    );
+
+    private final HttpServerTransport httpTransport;
+
+    public ReadinessService(ClusterService clusterService, Environment environment, HttpServerTransport httpTransport) {
+        this.httpTransport = httpTransport;
+        this.serverChannel = null;
+        this.environment = environment;
+        clusterService.addListener(this);
+
+        this.workerExecutor = EsExecutors.newScaling(
+            "readiness-worker",
+            0,
+            EsExecutors.allocatedProcessors(environment.settings()) * 2,
```

Comment:
Why do we need more than 1 thread for this? The thread should basically never be blocked, just return immediately right?

Reply:
Since I'm sending bytes back to the readiness-cli here I thought it might be slow on receiving sometimes and we'd block the only thread we have serving requests. I've removed the executor now since with the TCP probe method we just close the socket immediately.